features/steps/component.go

```diff
@@ -0,0 +1,235 @@
+	cfg              *config.Config
+	wg               *sync.WaitGroup
+	signals          chan os.Signal
+	waitEventTimeout time.Duration
+}
+
+func NewComponent(t *testing.T) *Component {
+	return &Component{
+		errorChan:        make(chan error),
+		svcStarted:       make(chan bool, 1),
+		DatasetAPI:       httpfake.New(httpfake.WithTesting(t)),
+		wg:               &sync.WaitGroup{},
+		waitEventTimeout: time.Second * 5,
+	}
+}
+
+// initService initialises the server, the mocks and waits for the dependencies to be ready
+func (c *Component) initService(ctx context.Context) error {
+	// register interrupt signals
+	c.signals = make(chan os.Signal, 1)
+	signal.Notify(c.signals, os.Interrupt, syscall.SIGTERM)
+
+	cfg, err := config.Get()
+	if err != nil {
+		return fmt.Errorf("failed to get config: %w", err)
+	}
+
+	log.Info(ctx, "config read", log.Data{"cfg": cfg})
+
+	cfg.DatasetAPIURL = c.DatasetAPI.ResolveURL("")
+
+	s3Config := &aws.Config{
+		Credentials:      credentials.NewStaticCredentials(cfg.MinioAccessKey, cfg.MinioSecretKey, ""),
+		Endpoint:         aws.String(cfg.LocalObjectStore),
+		Region:           aws.String(cfg.AWSRegion),
+		DisableSSL:       aws.Bool(true),
+		S3ForcePathStyle: aws.Bool(true),
+	}
+
+	s := session.New(s3Config)
+	c.S3Downloader = s3manager.NewDownloader(s)
+
+	// producer for triggering test events
+	if c.producer, err = kafka.NewProducer(
+		ctx,
+		cfg.Kafka.Addr,
+		cfg.Kafka.CantabularCSVCreatedTopic,
+		kafka.CreateProducerChannels(),
+		&kafka.ProducerConfig{
+			KafkaVersion:    &cfg.Kafka.Version,
+			MaxMessageBytes: &cfg.Kafka.MaxBytes,
+		},
+	); err != nil {
+		return fmt.Errorf("error creating kafka producer: %w", err)
+	}
+
+	// consumer for receiving events
+	cgChannels := kafka.CreateConsumerGroupChannels(1)
+	kafkaOffset := kafka.OffsetNewest
+	if cfg.Kafka.OffsetOldest {
+		kafkaOffset = kafka.OffsetOldest
+	}
+
+	if c.consumer, err = kafka.NewConsumerGroup(
+		ctx,
+		cfg.Kafka.Addr,
+		cfg.Kafka.CantabularCSVWCreatedTopic,
+		"csvw-created-group",
+		cgChannels,
+		&kafka.ConsumerGroupConfig{
+			KafkaVersion: &cfg.Kafka.Version,
+			Offset:       &kafkaOffset,
+		},
+	); err != nil {
+		return fmt.Errorf("error creating kafka consumer: %w", err)
+	}
+
+	// start kafka logging go-routines
+	c.producer.Channels().LogErrors(ctx, "component producer")
+	c.consumer.Channels().LogErrors(ctx, "component consumer")
+
+	// wait for consumer and producer to be ready
+	<-c.consumer.Channels().Ready
+	log.Info(ctx, "consumer ready")
+
+	<-c.producer.Channels().Ready
+	log.Info(ctx, "producer ready")
+
+	// Create service and initialise it
+	c.svc = service.New()
+	if err = c.svc.Init(ctx, cfg, BuildTime, GitCommit, Version); err != nil {
+		return fmt.Errorf("failed to init service: %w", err)
+	}
+
+	c.cfg = cfg
+
+	return nil
+}
+
+func (c *Component) startService(ctx context.Context) {
+	defer c.wg.Done()
+
+	c.svc.Start(context.Background(), c.errorChan)
+	c.svcStarted <- true
```

Comment:
good solution (instead of defining the service start as a separate step) - I thought the issue was that we were starting the service before some steps were defined, but maybe this wasn't the case?

Reply:
I figured it was the component tests running before the kafka consumer was initialised properly (which happens in the service startup)